pkg/logger/ecs_error_log.go

```diff
@@ -0,0 +1,250 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package logger
+
+import (
+	"fmt"
+	"log"
+	"os"
+	"path/filepath"
+	"strconv"
+	"time"
+
+	"go.uber.org/zap/zapcore"
+)
+
+// ECSErrorLogger is used to send the error level log to a separated file and rotate
+// error logs when expired, ECS agent will mount the filepath and read it periodically
+type ECSErrorLogger struct {
+	ErrorLogFilePath    string
+	ErrorLogMaxAge      int
+	ErrorLogFileMaxSize int
+
+	errorChannel chan zapcore.Entry
+	queue        []zapcore.Entry
+	ticker       time.Ticker
+	errorFile    *os.File
+	size         int
+}
+
+// NewECSErrorLogger could new a struct based on the 3 environment variables:
+// error file path, error log TTL and error file max size
+func NewECSErrorLogger() *ECSErrorLogger {
+	ecsErrorLogFilePath := os.Getenv("STATUS_MESSAGE_FILE_PATH")
+	errorLogMaxAge, err := strconv.Atoi(os.Getenv("STATUS_MESSAGE_TTL"))
+	if err != nil || errorLogMaxAge <= 0 {
+		log.Fatalf("failed to get STATUS_MESSAGE_TTL: %v", err)
+	}
+	errorFileSize, err := strconv.Atoi(os.Getenv("STATUS_MESSAGE_MAX_BYTE_LENGTH"))
+	if err != nil || errorFileSize <= 0 {
+		log.Fatalf("failed to get STATUS_MESSAGE_MAX_BYTE_LENGTH: %v", err)
+	}
+	if errorFileSize > 1000 {
+		log.Printf("error log file size should not exceed 1KB, will use 1000 instead")
+		errorFileSize = 1000
+	}
+	return &ECSErrorLogger{
+		ErrorLogFilePath:    ecsErrorLogFilePath,
+		ErrorLogMaxAge:      errorLogMaxAge,
+		ErrorLogFileMaxSize: errorFileSize,
+		errorChannel:        make(chan zapcore.Entry, 100),
+	}
+}
+
+// Run func help select the different case between either write error log into file or
+// rotate the old logs in file based on TTL
+func (l *ECSErrorLogger) Run() {
+	defer l.ticker.Stop()
+	for {
+		select {
+		case newError := <-l.errorChannel:
+			l.Write(newError)
+		case <-l.ticker.C:
+			l.processTimeout()
+		}
+	}
+}
+
+// Write function could write the new error into the file, when the file size plus new error size
+// exceed the max limit, will call rotate function to delete the old log
+func (l *ECSErrorLogger) Write(newError zapcore.Entry) {
+	if l.errorFile == nil {
+		if err := l.openExistingOrNewErrorFile(); err != nil {
+			log.Printf("[ecs error reporter] could not open error log file when write new error, err: %v", err)
+			return
+		}
+	}
+	newErrorLog := []byte(fmt.Sprintf("{%+v, Level:%+v, Caller:%+v, Message:%+v}\r\n",
+		newError.Time, newError.Level, newError.Caller, newError.Message))
+
+	if len(newErrorLog) > l.ErrorLogFileMaxSize {
+		log.Printf("[ecs error reporter] error size exceed the max size of error file")
+		return
+	}
+
+	err := l.errorFile.Truncate(0)
```

Comment:
Why is this truncating the log file on every error?

Reply:
since we need to go through and rotate all the errors if expired or exceed the file size

Reply:
How does this behave with a high rate of errors being logged?  What impact does it have on collector performance?  Will the consumer of this data even be likely to see any/valid/complete data if the new data aren't prepared until after the old data are removed?

Reply:
actually customer will not directly use the error file, just like in the description, ECS agent will mount the file path and read the file, we will have a performance testing with agent team next year after they implement their code part.

Reply:
If nothing is using the data written by this logger then we should take the time to take a step back and address the concerns that the current implementation presents.  We definitely shouldn't ship it before we have assessed its performance and behavior under stress.